lib/compilation_engine.rb

```diff
@@ -188,19 +188,29 @@ def compile_let
     advance # let
 
     variable_name = @tokenizer.identifier
+    segment = Utils.kind_to_segment(@symbol_table.kind_of(variable_name))
+    index = @symbol_table.index_of(variable_name)
     advance # varName
 
     if symbol_token?("[")
       advance # [
-      compile_expression
+      array_index = @expression_parser.parse_expression
       advance # ]
-    end
+      advance # =
+      compile_expression # expression
+      array_index.write_vm_code(@vm_writer, @symbol_table)
 
-    advance # =
-    compile_expression # expression
+      @vm_writer.write_push(segment, index)
+      @vm_writer.write_arithmetic(:ADD)
+      @vm_writer.write_pop(:POINTER, 1)
+      @vm_writer.write_pop(:THAT, 0)
 
-    kind = @symbol_table.kind_of(variable_name)
-    @vm_writer.write_pop(Utils.kind_to_segment(kind), @symbol_table.index_of(variable_name))
+    else
+      advance # =
+      compile_expression # expression
+
+      @vm_writer.write_pop(segment, index)
+    end
```

Comment:
When we paired we talked about ways of removing duplication from this code. You already moved `segment` and `index` out of the conditional, but here’s a way we could deduplicate the middle part (parsing the `=` and the value being assigned) by splitting the conditional into two parts:

```ruby
if symbol_token?("[")
  advance # [
  array_index = @expression_parser.parse_expression
  advance # ]
end

advance # =
compile_expression # expression

if array_index
  array_index.write_vm_code(@vm_writer, @symbol_table)

  @vm_writer.write_push(segment, index)
  @vm_writer.write_arithmetic(:ADD)
  @vm_writer.write_pop(:POINTER, 1)
  @vm_writer.write_pop(:THAT, 0)
else
  @vm_writer.write_pop(segment, index)
end
```

So that’s like: we remember the AST for the array index expression if there is one, then we emit the VM code for the value being assigned, then we either store that value in an array (if we have an array index AST) or in the variable directly (if we don’t).

WDYT?